### PR TITLE
[#1533] Flexa integration

### LINKED ISSUE #1533: Flexa Integration
- integrate Flexa's Layers SDK + SpendKit
- GitHub Docs for Layers Android SDK are in Flexa's GitHub repo that has been shared with your business email
- Slack channel for support: https://zcash.slack.com/archives/C06VAKMB58E
- API key: https://zcash.slack.com/archives/C06VAKMB58E/p1713884591419739?thread_ts=1713820382.855949&cid=C06VAKMB58E

### PR DESCRIPTION
**Pull request by Milan-Cerovsky**
Closes #1533

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [x] Update **documentation** as appropriate (e.g [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), [**CHANGELOG.md**](../blob/main/CHANGELOG.md), etc.)
- [x] **Run the app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [x] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [x] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [x] Review the **documentation**, [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), etc. as appropriate
- [x] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -135,6 +135,7 @@ dependencyResolutionManagement {
                 }
             }
         }
+        mavenLocal()
```
Comment (HonzaR) [inline settings.gradle.kts:138]: Why do we need this change, please?

File: ui-lib/build.gradle.kts [hunk 2]
```diff
@@ -122,6 +131,9 @@ dependencies {
     implementation(libs.zcash.bip39)
     implementation(libs.zxing)
 
+    api("com.flexa:core:1.0.1")
+    api("com.flexa:spend:1.0.1")
```
Comment (HonzaR) [inline ui-lib/build.gradle.kts:135]: Can we use the same dependency approach as we do elsewhere, injecting them via `settings.gradle`?

File: preference-impl-android-lib/src/main/java/co/electriccoin/zcash/preference/InstallationPreferenceProvider.kt
Comment (HonzaR) [inline preference-impl-android-lib/src/main/java/co/electriccoin/zcash/preference/InstallationPreferenceProvider.kt:1]: What's the reason for creating a new preference provider and not using the existing one, e.g., `EncryptedPreferenceProvider`?

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/BalanceRepository.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.WhileSubscribed
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.stateIn
+
+interface BalanceRepository {
+    /**
+     * A flow of the wallet balances state used for the UI layer. It's computed form [WalletSnapshot]'s properties
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/BalanceRepository.kt:23]: Typo `form` -> `from`

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                        "        }\n" +
+                        "    }\n" +
+                        "}"
+            )
+        }
+
+    private suspend fun createFlexaAccount(zecBalance: Double) =
+        AppAccount(
+            accountId = getFlexaAccountId(),
+            displayName = "My Wallet",
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt:95]: Is this our preferred `displayName` value for each user?

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/home/view/HomeView.kt [hunk 5]
```diff
@@ -112,7 +112,6 @@ private fun HomeContent(
             key = { index ->
                 subScreens[index].title
             },
-            beyondBoundsPageCount = 1,
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/home/view/HomeView.kt:115]: What is the intention behind this change, please?

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            backNavigationCommand.emit(Unit)
+        }
+
+    private fun onBuyWithCoinbaseClicked() =
+        viewModelScope.launch {
+            val appId = BuildConfig.ZCASH_COINBASE_APP_ID
+
+            when {
+                appId.isEmpty() && isDebug ->
+                    coinbaseNavigationCommand.emit("https://www.coinbase.com") // fallback debug url
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt:89]: Can we move these hardcoded urls to constants?

File: ui-lib/src/main/res/ui/integrations/values/strings.xml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+    <string name="integrations_title">Integrations</string>
+    <string name="integrations_coinbase">Buy <xliff:g id="currency" example="ZEC">%1$s</xliff:g> with Coinbase</string>
+    <string name="integrations_coinbase_subtitle">A hassle-free way to buy <xliff:g id="currency" example="ZEC">%1$s</xliff:g> and get it directly into your Zashi wallet.</string>
+    <string name="integrations_flexa">Pay with Flexa</string>
+    <string name="integrations_flexa_subtitle">Pay with Flexa payment clips and explore a new way of spending Zcash.</string>
+    <string name="integrations_version">Version %s</string>
```
Comment (HonzaR) [inline ui-lib/src/main/res/ui/integrations/values/strings.xml:7]: Let's use xliff instead here.

**Review by HonzaR — COMMENTED** [review]
Thank you for these changes! I'm sharing a few first ideas in the comments.

File: settings.gradle.kts (hunk 1 above)
Comment (Milan-Cerovsky) [reply settings.gradle.kts:138]: flexa is not on maven, I am instead building their sdk locally lol

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt (hunk 4 above)
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt:95]: I dont know yet :D

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/BiometricActivity.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                    override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
+                        super.onAuthenticationSucceeded(result)
+                        biometricRepository.onBiometricResult(BiometricResult.Success(requestCode))
+                        finish()
+                    }
+
+                    override fun onAuthenticationFailed() {
+                        super.onAuthenticationFailed()
+                        biometricRepository.onBiometricResult(BiometricResult.Failure(requestCode))
+                        finish()
```
Comment (HonzaR) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/BiometricActivity.kt:45]: We should not finish the activity here, as the failed state can be caused, e.g., by a wrong pin set or an unsupported finger used in the biometric authentication.

**Review by HonzaR — CHANGES_REQUESTED** [review]

File: ui-lib/build.gradle.kts (hunk 2 above)
Comment (Milan-Cerovsky) [reply ui-lib/build.gradle.kts]: fixed

File: preference-impl-android-lib/src/main/java/co/electriccoin/zcash/preference/InstallationPreferenceProvider.kt
Comment (Milan-Cerovsky) [reply preference-impl-android-lib/src/main/java/co/electriccoin/zcash/preference/InstallationPreferenceProvider.kt:1]: deleted

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/BiometricActivity.kt (hunk 8 above)
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/BiometricActivity.kt:39]: fixed

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt (hunk 4 above)
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/FlexaRepository.kt:97]: done

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt (hunk 6 above)
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt]: no need to use constants here, especially for the formatted url below :) they are used only in 1 place

File: ui-lib/src/main/res/ui/integrations/values/strings.xml (hunk 7 above)
Comment (Milan-Cerovsky) [reply ui-lib/src/main/res/ui/integrations/values/strings.xml:7]: done

File: ui-lib/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -22,6 +22,10 @@
             android:label="@string/app_name"
             android:theme="@style/Theme.App.Starting" />
 
+        <activity
+            android:name=".BiometricActivity"
+            android:theme="@style/Theme.App.Transparent" />
```
[the remark this answers is not in the document]
Comment (Milan-Cerovsky) [reply ui-lib/src/main/AndroidManifest.xml:27]: done

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt
[the remark this answers is not in the document]
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt:1]: we should cancel the transaction probably

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt [hunk 10]
```diff
@@ -98,4 +142,175 @@ class IntegrationsViewModel(
...
+                        Twig.debug { "Transaction successful $result" }
+                        Flexa.buildSpend()
+                            .transactionSent(
+                                commerceSessionId = transaction.getOrNull()?.commerceSessionId.orEmpty(),
+                                txSignature = result.second.orEmpty()
+                            )
+                    }
+
+                    else -> {
+                        Twig.error { "Transaction submission failed" }
```
[the remark this answers is not in the document]
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt:177]: we should cancel the transaction probably

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt
[the remark this answers is not in the document]
Comment (Milan-Cerovsky) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt:1]: there were slight changes, especially changing the biometrics, if we will do this in the future somewhere else we might put this whole thing into usecase

File: build.gradle.kts [hunk 11]
```diff
@@ -45,6 +45,7 @@ buildscript {
                 }
             }
         }
+        maven("${rootProject.projectDir}/maven")
```
[the remark this answers is not in the document]
Comment (Milan-Cerovsky) [reply build.gradle.kts]: done

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (HonzaR) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/screen/integrations/viewmodel/IntegrationsViewModel.kt:175]: Yes, let's cancel the transaction here once we have the necessary cancellation API from Flexa.

**Review by HonzaR — APPROVED** [review]
Approving with several comments inline.

**Pull request merged**